# Welcome to Maji Ndogo
The story we'll step into is not unique to Maji Ndogo; it mirrors real-world challenges faced in many places across the globe.
Our mission is to help rejuvenate the drying Mto wa Matumaini – The River of Hope – using a data-driven approach. 


# Integrated Project 1
Part 1: Beginning our data-driven journey in Maji Ndogo.
At the end of this project, we will be able to:

1. Explore a realistic database with SQL. 

2. Use SQL to clean and explore a database with 60,000 unique records. 

3. Become immersed in a story-driven simulation of real data projects, data analysis, and good leadership.

Congratulations on the successful completion of our extensive survey - a database of 60,000 records, meticulously collected by our devoted team of engineers, field workers, scientists, and analysts. Now, the next crucial phase of our mission begins. We need to make sense of this immense data trove and extract meaningful insights. We must breathe life into these records and listen to the story they are telling us.

First, we have to load our database.

In [ ]:
%reload_ext sql

In [60]:
%sql mysql+pymysql://root:Pk_0270197907@localhost:3306/md_water_services

Connecting and switching to connection 'mysql+pymysql://root:***@localhost:3306/md_water_services'

# 1. Get to know our data:
Start by retrieving the first few records from each table. How many tables are there in our database? What are the names of these tables? 


In [61]:
%%sql
SHOW TABLES;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

8 rows affected.

+-----------------------------+
| Tables_in_md_water_services |
+-----------------------------+
|       data_dictionary       |
|           employee          |
|     global_water_access     |
|           location          |
|            visits           |
|        water_quality        |
|         water_source        |
|        well_pollution       |
+-----------------------------+

In [40]:
%config SqlMagic.displaylimit = 10

So let's have a look at one of these tables, Let's use location.

In [62]:
%%sql
SELECT
    *
FROM
    location
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+-------------+----------------------+---------------+-----------+---------------+
| location_id |       address        | province_name | town_name | location_type |
+-------------+----------------------+---------------+-----------+---------------+
|  AkHa00000  |  2 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00001  | 10 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00002  |  9 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00003  | 139 Addis Ababa Road |     Akatsi    |   Harare  |     Urban     |
|  AkHa00004  | 17 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
+-------------+----------------------+---------------+-----------+---------------+

So we can see that this table has information on a specific location, with an address, the province and town the location is in, and if it's in a city (Urban) or not. We can't really see what location this is but we can see some sort of identifying number of that location.

Ok, so let's look at the visits table

In [63]:
%%sql
SELECT
    *
FROM
    visits;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

60146 rows affected.

+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
| record_id | location_id |  source_id   |    time_of_record   | visit_count | time_in_queue | assigned_employee_id |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
|     0     |  SoIl32582  | SoIl32582224 | 2021-01-01 09:10:00 |      1      |       15      |          12          |
|     1     |  KiRu28935  | KiRu28935224 | 2021-01-01 09:17:00 |      1      |       0       |          46          |
|     2     |  HaRu19752  | HaRu19752224 | 2021-01-01 09:36:00 |      1      |       62      |          40          |
|     3     |  AkLu01628  | AkLu01628224 | 2021-01-01 09:53:00 |      1      |       0       |          1           |
|     4     |  AkRu03357  | AkRu03357224 | 2021-01-01 10:11:00 |      1      |       28      |          14          |
|     5     |  KiRu29315  | KiRu29315224 | 2021-01-01 10:17:00 |      1      |       9       |          40          |
|     6     |  AkRu05234  | AkRu05234224 | 2021-01-01 10:18:00 |      1      |       0       |          30          |
|     7     |  KiRu28520  | KiRu28520224 | 2021-01-01 10:28:00 |      1      |       0       |          34          |
|     8     |  HaZa21742  | HaZa21742224 | 2021-01-01 10:37:00 |      1      |       0       |          6           |
|     9     |  AmDa12214  | AmDa12214224 | 2021-01-01 10:58:00 |      1      |       0       |          36          |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
Truncated to displaylimit of 10.

So this is a list of location_id, source_id, record_id, and a date and time, so it makes sense that someone (assigned_employee_id) visited some location (location_id) at some time (time_of_record ) and found a 'source' there (source_id). Often the "_id" columns are related to another table. In this case, the source_id in the visits table refers to source_id in the water_source table. This is what we call a foreign key.

So let's look at the water_source table to see what a 'source' is.

In [64]:
%%sql
SELECT
    *
FROM
    water_source;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

39650 rows affected.

+--------------+----------------------+-------------------------+
|  source_id   | type_of_water_source | number_of_people_served |
+--------------+----------------------+-------------------------+
| AkHa00000224 |     tap_in_home      |           956           |
| AkHa00001224 |  tap_in_home_broken  |           930           |
| AkHa00002224 |  tap_in_home_broken  |           486           |
| AkHa00003224 |         well         |           364           |
| AkHa00004224 |  tap_in_home_broken  |           942           |
| AkHa00005224 |     tap_in_home      |           736           |
| AkHa00006224 |     tap_in_home      |           882           |
| AkHa00007224 |     tap_in_home      |           554           |
| AkHa00008224 |         well         |           398           |
| AkHa00009224 |         well         |           346           |
+--------------+----------------------+-------------------------+
Truncated to displaylimit of 10.

Nice! Ok, we're getting somewhere now... Water sources are where people get their water from! Ok, this database is actually complex,so maybe a good idea for you is to look at the rest of the tables quickly. You can just select them, but remember in good SQL there would be a data dictionary somewhere that documents all of this information, so you should read that as well.
A data dictionary has been embedded into the database. If you query the data_dictionary table, an explanation of each column is
given there.

In [65]:
%%sql
SELECT
    *
FROM
    data_dictionary;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

49 rows affected.

+------------+----------------------+-------------------------------------------------+--------------+-----------------------------+
| table_name |     column_name      |                   description                   |   datatype   |          related_to         |
+------------+----------------------+-------------------------------------------------+--------------+-----------------------------+
|  employee  | assigned_employee_id |       Unique ID assigned to each employee       |     INT      |            visits           |
|  employee  |    employee_name     |               Name of the employee              | VARCHAR(255) |                             |
|  employee  |     phone_number     |          Contact number of the employee         | VARCHAR(15)  |                             |
|  employee  |        email         |          Email address of the employee          | VARCHAR(255) |                             |
|  employee  |       address        |       Residential address of the employee       | VARCHAR(255) |                             |
|  employee  |      town_name       |   Name of the town where the employee resides   | VARCHAR(255) |                             |
|  employee  |    province_name     | Name of the province where the employee resides | VARCHAR(255) |                             |
|  employee  |       position       |      Position or job title of the employee      | VARCHAR(255) |                             |
|   visits   |      record_id       |         Unique ID assigned to each visit        |     int      | water_quality, water_source |
|   visits   |     location_id      |            ID of the location visited           | varchar(255) |           location          |
+------------+----------------------+-------------------------------------------------+--------------+-----------------------------+
Truncated to displaylimit of 10.

# 2. Dive into the water sources:
Now that you're familiar with the structure of the tables, let's dive deeper. We need to understand the types of water sources we're dealing with.

In [66]:
%%sql
SELECT
    DISTINCT(type_of_water_source)
FROM
    water_source;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+
| type_of_water_source |
+----------------------+
|     tap_in_home      |
|  tap_in_home_broken  |
|         well         |
|      shared_tap      |
|        river         |
+----------------------+

Let me quickly bring you up to speed on these water source types:

1. River - People collect drinking water along a river. This is an open water source that millions of people use in Maji Ndogo. Water from a river has a high risk of being contaminated with biological and other pollutants, so it is the worst source of water possible.

2. Well - These sources draw water from underground sources, and are commonly shared by communities. Since these are closed water sources, contamination is much less likely compared to a river. Unfortunately, due to the aging infrastructure and the corruption of officials in the past, many of our wells are not clean.

3. Shared tap - This is a tap in a public area shared by communities.

4. Tap in home - These are taps that are inside the homes of our citizens. On average about 6 people live together in Maji Ndogo, so each of these taps serves about 6 people.

5. Broken tap in home - These are taps that have been installed in a citizen’s home, but the infrastructure connected to that tap is  not functional. This can be due to burst pipes, broken pumps or water treatment plants that are not working.

An important note on the home taps: About 6-10 million people have running water installed in their homes in Maji Ndogo, including broken taps. If we were to document this, we would have a row of data for each home, so that one record is one tap. That means our database would contain about 1 million rows of data, which may slow our systems down. For now, the surveyors combined the data of many households together into a single record.

For example, the first record, AkHa00000224 is for a tap_in_home that serves 956 people. What this means is that the records of about 160 homes nearby were combined into one record, with an average of 6 people living in each house 160 x 6 ≈ 956. So 1 tap_in_home or tap_in_home_broken record actually refers to multiple households, with the sum of the people living in these homes equal to number_of_people_served.

# 3. Unpack the visits to water sources:
We have a table in our database that logs the visits made to different water sources.Let's write an SQL query that retrieves all records from this table where the time_in_queue is more than some crazy time, say 500 min. How
would it feel to queue 8 hours for water?

In [67]:
%%sql
SELECT
    *
FROM
    visits
WHERE
    time_in_queue > 500;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

105 rows affected.

+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
| record_id | location_id |  source_id   |    time_of_record   | visit_count | time_in_queue | assigned_employee_id |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
|    899    |  SoRu35083  | SoRu35083224 | 2021-01-16 10:14:00 |      6      |      515      |          28          |
|    2304   |  SoKo33124  | SoKo33124224 | 2021-02-06 07:53:00 |      5      |      512      |          16          |
|    2315   |  KiRu26095  | KiRu26095224 | 2021-02-06 14:32:00 |      3      |      529      |          8           |
|    3206   |  SoRu38776  | SoRu38776224 | 2021-02-20 15:03:00 |      5      |      509      |          46          |
|    3701   |  HaRu19601  | HaRu19601224 | 2021-02-27 12:53:00 |      3      |      504      |          0           |
|    4154   |  SoRu38869  | SoRu38869224 | 2021-03-06 10:44:00 |      2      |      533      |          24          |
|    5483   |  AmRu14089  | AmRu14089224 | 2021-03-27 18:15:00 |      4      |      509      |          12          |
|    9177   |  SoRu37635  | SoRu37635224 | 2021-05-22 18:48:00 |      2      |      515      |          1           |
|    9648   |  SoRu36096  | SoRu36096224 | 2021-05-29 11:24:00 |      2      |      533      |          3           |
|   11631   |  AkKi00881  | AkKi00881224 | 2021-06-26 06:15:00 |      6      |      502      |          32          |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
Truncated to displaylimit of 10.

What type of water sources take this long to queue for. We will have to find that information in another table that lists the types of water sources. So let's write down a few of these source_id values from our results, and search for them in the other table.

I chose these:

SoRu35083224
KiRu26095224
HaRu19601224

In [68]:
%%sql
SELECT
    *
FROM
    water_source
WHERE
    source_ID IN ('SoRu35083224', 'KiRu26095224', 'HaRu19601224');

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

3 rows affected.

+--------------+----------------------+-------------------------+
|  source_id   | type_of_water_source | number_of_people_served |
+--------------+----------------------+-------------------------+
| HaRu19601224 |      shared_tap      |           3322          |
| KiRu26095224 |      shared_tap      |           3734          |
| SoRu35083224 |      shared_tap      |           3058          |
+--------------+----------------------+-------------------------+

# 4. Assess the quality of water sources:
The quality of our water sources is the whole point of this survey. We have a table that contains a quality score for each visit made about a water source that was assigned by a Field surveyor. They assigned a score to each source from 1, being terrible, to 10 for a good, clean water source in a home. Shared taps are not rated as high, and the score also depends on how long the queue times are.

In [45]:
%%sql
SELECT
    *
FROM
    water_quality;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

60146 rows affected.

+-----------+--------------------------+-------------+
| record_id | subjective_quality_score | visit_count |
+-----------+--------------------------+-------------+
|     0     |            0             |      1      |
|     1     |            1             |      1      |
|     2     |            5             |      1      |
|     3     |            10            |      1      |
|     4     |            4             |      1      |
|     5     |            0             |      1      |
|     6     |            9             |      1      |
|     7     |            10            |      1      |
|     8     |            2             |      1      |
|     9     |            10            |      1      |
+-----------+--------------------------+-------------+
Truncated to displaylimit of 10.

Let's check if this is true. The surveyors only made multiple visits to shared taps and did not revisit other types of water sources. So there should be no records of second visits to locations where there are good water sources, like taps in homes.

So let's write a query to find records where the subject_quality_score is 10 -- only looking for home taps -- and where the source was visited a second time.

In [69]:
%%sql
SELECT
    *
FROM
    water_quality
WHERE
    subjective_quality_score = 10
AND
    visit_count = 2;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

218 rows affected.

+-----------+--------------------------+-------------+
| record_id | subjective_quality_score | visit_count |
+-----------+--------------------------+-------------+
|     59    |            10            |      2      |
|    137    |            10            |      2      |
|    269    |            10            |      2      |
|    363    |            10            |      2      |
|    378    |            10            |      2      |
|    618    |            10            |      2      |
|    752    |            10            |      2      |
|    801    |            10            |      2      |
|    819    |            10            |      2      |
|    850    |            10            |      2      |
+-----------+--------------------------+-------------+
Truncated to displaylimit of 10.

This should not be happening. I think some of our employees may have made mistakes. We can appoint an Auditor to check some of the data independently, and make sure we have the right information.

# 5. Investigate pollution issues:
Did you notice that we recorded contamination/pollution data for all of the well sources? Let's find the right table and print the first few rows.

In [70]:
%%sql
SELECT
    *
FROM
    well_pollution;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17383 rows affected.

+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+
|  source_id   |         date        |                       description                        | pollutant_ppm |  biological |         results          |
+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+
| KiRu28935224 | 2021-01-04 09:17:00 |                Bacteria: Giardia Lamblia                 |      0.0      |   495.898   | Contaminated: Biological |
| AkLu01628224 | 2021-01-04 09:53:00 |                    Bacteria: E. coli                     |      0.0      |   6.09608   | Contaminated: Biological |
| HaZa21742224 | 2021-01-04 10:37:00 |    Inorganic contaminants: Zinc, Zinc, Lead, Cadmium     |     2.715     |     0.0     |  Contaminated: Chemical  |
| HaRu19725224 | 2021-01-04 11:04:00 |                          Clean                           |   0.0288593   | 9.56996e-05 |          Clean           |
| SoRu35703224 | 2021-01-04 11:29:00 |                    Bacteria: E. coli                     |      0.0      |   22.5009   | Contaminated: Biological |
| AkHa00070224 | 2021-01-04 11:42:00 |             Inorganic contaminants: Cadmium              |    5.46739    |     0.0     |  Contaminated: Chemical  |
| HaSe21346224 | 2021-01-04 11:52:00 |                          Clean                           |   0.0140376   | 8.98989e-05 |          Clean           |
| HaYa21468224 | 2021-01-04 12:03:00 | Inorganic contaminants: Chromium, Barium, Chromium, Lead |    6.05137    |     0.0     |  Contaminated: Chemical  |
| SoRu36278224 | 2021-01-04 12:24:00 |                Parasite: Cryptosporidium                 |      0.0      |   485.162   | Contaminated: Biological |
| AkLu02155224 | 2021-01-04 12:29:00 |        Inorganic contaminants: Selenium, Arsenic         |    7.64106    |     0.0     |  Contaminated: Chemical  |
+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+
Truncated to displaylimit of 10.

It looks like our scientists diligently recorded the water quality of all the wells. Some are contaminated with biological contaminants, while others are polluted with an excess of heavy metals and other pollutants. Based on the results, each well was classified as Clean, Contaminated: Biological or Contaminated: Chemical. It is important to know this because wells that are polluted with bio- or other contaminants are not safe to drink. It looks like they recorded the source_id of each test, so we can link it to a source, at some place in Maji Ndogo.

In the well pollution table, the descriptions are notes taken by our scientists as text, so it will be challenging to process it. The biological column is in units of CFU/mL, so it measures how much contamination is in the water. 0 is clean, and anything more than 0.01 is contaminated. Let's check the integrity of the data. The worst case is if we have contamination, but we think we don't. People can get sick, so we need to make sure there are no errors here.

In [71]:
%%sql
SELECT
    *
FROM
    well_pollution
WHERE
    results = "clean"
AND
    biological > 0.01;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

64 rows affected.

+--------------+---------------------+---------------------------------+---------------+------------+---------+
|  source_id   |         date        |           description           | pollutant_ppm | biological | results |
+--------------+---------------------+---------------------------------+---------------+------------+---------+
| AkRu08936224 | 2021-01-08 09:22:00 |        Bacteria: E. coli        |   0.0406458   |  35.0068   |  Clean  |
| AkRu06489224 | 2021-01-10 09:44:00 | Clean Bacteria: Giardia Lamblia |   0.0897904   |   38.467   |  Clean  |
| SoRu38011224 | 2021-01-14 15:35:00 |        Bacteria: E. coli        |   0.0425095   |  19.2897   |  Clean  |
| AkKi00955224 | 2021-01-22 12:47:00 |        Bacteria: E. coli        |   0.0812092   |  40.2273   |  Clean  |
| KiHa22929224 | 2021-02-06 13:54:00 |        Bacteria: E. coli        |   0.0722537   |  18.4482   |  Clean  |
| KiRu25473224 | 2021-02-07 15:51:00 | Clean Bacteria: Giardia Lamblia |   0.0630094   |  24.4536   |  Clean  |
| HaRu17401224 | 2021-03-01 13:44:00 | Clean Bacteria: Giardia Lamblia |   0.0649209   |  25.8129   |  Clean  |
| AkRu07137224 | 2021-03-04 13:41:00 | Clean Bacteria: Giardia Lamblia |   0.0656843   |  18.2978   |  Clean  |
| KiRu27205224 | 2021-03-13 14:17:00 | Clean Bacteria: Giardia Lamblia |   0.0418018   |  49.4281   |  Clean  |
| AkLu02307224 | 2021-03-13 15:41:00 |        Bacteria: E. coli        |   0.0709682   |   35.203   |  Clean  |
+--------------+---------------------+---------------------------------+---------------+------------+---------+
Truncated to displaylimit of 10.

If we compare the results of this query to the entire table it seems like we have some inconsistencies in how the well statuses are recorded. Specifically, it seems that some data input personnel might have mistaken the description field for determining the cleanliness of the water.

It seems like, in some cases, if the description field begins with the word “Clean”, the results have been classified as “Clean” in the results column, even though the biological column is > 0.0

When we work with real-world data we may find inconsistencies due to data being misinterpreted based on a description rather than its actual values. Let’s dive deeper into the cause of the issue with the biological contamination data.

Vuyisile has told me that the descriptions should only have the word “Clean” if there is no biological contamination (and no chemical pollutants). Some data personnel must have copied the data from the scientist's notes into our database incorrectly. We need to find and remove the “Clean” part from all the descriptions that do have a biological contamination so this mistake is not made again. 

The second issue has arisen from this error, but it is much more problematic. Some of the field surveyors have marked wells as Clean in the results column because the description had the word “Clean” in it, even though they have a biological contamination. So we need to find all the results that have a value greater than 0.01 in the biological column and have been set to Clean in the results column. 

First, let's look at the descriptions. We need to identify the records that mistakenly have the word Clean in the description. However, it is important to remember that not all of our field surveyors used the description to set the results – some checked the actual data.

In [76]:
%%sql
SELECT
    *
FROM
    well_pollution
WHERE
    description LIKE 'Clean %';

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

38 rows affected.

+--------------+---------------------+---------------------------------+---------------+------------+---------+
|  source_id   |         date        |           description           | pollutant_ppm | biological | results |
+--------------+---------------------+---------------------------------+---------------+------------+---------+
| AkRu06489224 | 2021-01-10 09:44:00 | Clean Bacteria: Giardia Lamblia |   0.0897904   |   38.467   |  Clean  |
| KiRu25473224 | 2021-02-07 15:51:00 | Clean Bacteria: Giardia Lamblia |   0.0630094   |  24.4536   |  Clean  |
| HaRu17401224 | 2021-03-01 13:44:00 | Clean Bacteria: Giardia Lamblia |   0.0649209   |  25.8129   |  Clean  |
| AkRu07137224 | 2021-03-04 13:41:00 | Clean Bacteria: Giardia Lamblia |   0.0656843   |  18.2978   |  Clean  |
| KiRu27205224 | 2021-03-13 14:17:00 | Clean Bacteria: Giardia Lamblia |   0.0418018   |  49.4281   |  Clean  |
| AkHa00514224 | 2021-04-11 12:11:00 | Clean Bacteria: Giardia Lamblia |   0.0305404   |  22.0255   |  Clean  |
| AmAm09776224 | 2021-05-23 11:28:00 | Clean Bacteria: Giardia Lamblia |   0.0963821   |  13.6574   |  Clean  |
| SoIl32894224 | 2021-07-11 11:37:00 | Clean Bacteria: Giardia Lamblia |   0.0712408   |  5.44957   |  Clean  |
| AkRu07366224 | 2021-07-23 11:19:00 | Clean Bacteria: Giardia Lamblia |   0.0969458   |  26.0308   |  Clean  |
| KiHa23443224 | 2021-09-05 12:34:00 | Clean Bacteria: Giardia Lamblia |     0.0828    |  13.7162   |  Clean  |
| AkRu06340224 | 2021-11-01 15:24:00 | Clean Bacteria: Giardia Lamblia |    0.051746   |  36.4594   |  Clean  |
| AkRu06363224 | 2021-11-15 13:36:00 | Clean Bacteria: Giardia Lamblia |   0.0882158   |  45.3731   |  Clean  |
| AkRu08749224 | 2021-11-19 13:37:00 |     Clean Bacteria: E. coli     |   0.0888194   |  9.28085   |  Clean  |
| AkRu07066224 | 2021-12-10 14:40:00 |     Clean Bacteria: E. coli     |   0.0151035   |  44.4652   |  Clean  |
| KiAm21996224 | 2021-12-16 10:05:00 |     Clean Bacteria: E. coli     |   0.0480415   |  25.4174   |  Clean  |
| AkRu05377224 | 2022-01-01 10:53:00 |     Clean Bacteria: E. coli     |   0.0226614   |  0.613503  |  Clean  |
| HaRu20765224 | 2022-02-07 12:39:00 |     Clean Bacteria: E. coli     |   0.0213831   |  6.95778   |  Clean  |
| SoMa34473224 | 2022-03-10 09:01:00 |     Clean Bacteria: E. coli     |   0.0745784   |  13.6458   |  Clean  |
| SoKo33474224 | 2022-03-20 16:07:00 |     Clean Bacteria: E. coli     |   0.0687177   |  10.5712   |  Clean  |
| AkHa00226224 | 2022-03-26 11:53:00 |     Clean Bacteria: E. coli     |   0.0424571   |  39.4978   |  Clean  |
| AkRu06881224 | 2022-03-28 13:37:00 |     Clean Bacteria: E. coli     |   0.0202831   |  33.0692   |  Clean  |
| AkLu02254224 | 2022-04-25 12:51:00 |     Clean Bacteria: E. coli     |   0.0635533   |  16.9482   |  Clean  |
| AkRu08416224 | 2022-05-29 15:59:00 |     Clean Bacteria: E. coli     |   0.0406701   |  27.0805   |  Clean  |
| SoMa34205224 | 2022-07-31 13:25:00 |     Clean Bacteria: E. coli     |   0.0726217   |   23.892   |  Clean  |
| KiIs23482224 | 2022-08-15 16:41:00 |     Clean Bacteria: E. coli     |   0.0539137   |  41.0842   |  Clean  |
| AmRu13252224 | 2022-08-20 14:23:00 |     Clean Bacteria: E. coli     |   0.0653589   |  30.8643   |  Clean  |
| KiRu27297224 | 2022-11-21 13:18:00 |     Clean Bacteria: E. coli     |   0.0856842   |  8.62509   |  Clean  |
| KiRu26929224 | 2022-12-18 10:52:00 |     Clean Bacteria: E. coli     |   0.0748808   |  26.6541   |  Clean  |
| HaAm16042224 | 2022-12-22 13:52:00 |     Clean Bacteria: E. coli     |   0.0605767   |  4.54755   |  Clean  |
| AkRu03107224 | 2023-01-01 15:37:00 |     Clean Bacteria: E. coli     |   0.0901101   |  18.3118   |  Clean  |
| KiHa23111224 | 2023-01-12 15:56:00 |     Clean Bacteria: E. coli     |   0.0161869   |  36.8748   |  Clean  |
| AkRu03771224 | 2023-02-26 12:40:00 |     Clean Bacteria: E. coli     |   0.0625017   |  19.7648   |  Clean  |
| AkRu04253224 | 2023-02-26 13:46:00 |     Clean Bacteria: E. coli     |   0.027

In [75]:
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

displaylimit: Value None will be treated as 0 (no limit)

Looking at the results we can see two different descriptions that we need to fix:
1. All records that mistakenly have Clean Bacteria: E. coli should updated to Bacteria: E. coli
2. All records that mistakenly have Clean Bacteria: Giardia Lamblia should updated to Bacteria: Giardia Lamblia

The second issue we need to fix is in our results column. We need to update the results column from Clean to Contaminated: Biological where the biological column has a value greater than 0.01.


Now, when we change any data on the database, we need to be SURE there are no errors, as this could fill the database with incorrect values. A safer way to do the UPDATE is by testing the changes on a copy of the table first.

The CREATE TABLE new_table AS (query) approach is a neat trick that allows you to create a new table from the results set of a query.This method is especially useful for creating backup tables or subsets without the need for a separate CREATE TABLE and INSERT INTO statement.

In [73]:
%%sql
CREATE TABLE
    md_water_services.well_pollution_copy
AS (
    SELECT
        *
    FROM
        md_water_services.well_pollution
);

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17383 rows affected.

++
||
++
++

In [74]:
%%sql
UPDATE
    well_pollution_copy
SET
    description = 'Bacteria: E. coli'
WHERE
    description = 'Clean Bacteria: E. coli';

UPDATE
    well_pollution_copy
SET
    description = 'Bacteria: Giardia Lamblia'
WHERE
    description = 'Clean Bacteria: Giardia Lamblia';

UPDATE
    well_pollution_copy
SET
    results = 'Contaminated: Biological'
WHERE
    biological > 0.01 AND results = 'clean';

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

26 rows affected.

12 rows affected.

64 rows affected.

++
||
++
++

We can then check if our errors are fixed using a SELECT query on the well_pollution_copy table:

In [77]:
%%sql
SELECT
    *
FROM
    well_pollution_copy
WHERE
    description LIKE "Clean_%"
    OR (results = "Clean" AND biological > 0.01);

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

+-----------+------+-------------+---------------+------------+---------+
| source_id | date | description | pollutant_ppm | biological | results |
+-----------+------+-------------+---------------+------------+---------+
+-----------+------+-------------+---------------+------------+---------+